src/hotspot/share/services/stathist.cpp

```diff
@@ -0,0 +1,1192 @@
+#include "utilities/ostream.hpp"
+
+
+#include <locale.h>
+#include <time.h>
+
+
+// Define this switch to limit sampling to those values which can be obtained without locking.
+#undef NEVER_LOCK_WHEN_SAMPLING
+
+namespace StatisticsHistory {
+
+namespace counters {
+
+// These are counters for the statistics history. Ideally, they would live
+// inside their thematical homes, e.g. thread.cpp or classLoaderDataGraph.cpp,
+// however since this is unlikely ever to be brought upstream we keep them separate
+// from central coding to ease maintenance.
+static volatile size_t g_classes_loaded = 0;
+static volatile size_t g_classes_unloaded = 0;
+static volatile size_t g_threads_created = 0;
+
+void inc_classes_loaded(size_t count) {
+  Atomic::add(count, &g_classes_loaded);
+}
+
+void inc_classes_unloaded(size_t count) {
+  Atomic::add(count, &g_classes_unloaded);
+}
+
+void inc_threads_created(size_t count) {
+  Atomic::add(count, &g_threads_created);
+}
+
+} // namespace counters
+
+// helper function for the missing outputStream::put(int c, int repeat)
+static void ostream_put_n(outputStream* st, int c, int repeat) {
+  for (int i = 0; i < repeat; i ++) {
+    st->put(c);
+  }
+}
+
+static size_t record_size_in_bytes() {
+  const int num_columns = ColumnList::the_list()->num_columns();
+  return sizeof(record_t) + sizeof(value_t) * (num_columns - 1);
+}
+
+static void print_text_with_dashes(outputStream* st, const char* text, int width) {
+  assert(width > 0, "Sanity");
+  // Print the name centered within the width like this
+  // ----- system ------
+  int extra_space = width - (int)strlen(text);
+  if (extra_space > 0) {
+    int left_space = extra_space / 2;
+    int right_space = extra_space - left_space;
+    ostream_put_n(st, '-', left_space);
+    st->print_raw(text);
+    ostream_put_n(st, '-', right_space);
+  } else {
+    ostream_put_n(st, '-', width);
+  }
+}
+
+// Helper function for printing:
+// Print to ostream, but only if ostream is given. In any case return number ofcat_process = 10,
+// characters printed (or which would have been printed).
+static
+ATTRIBUTE_PRINTF(2, 3)
+int printf_helper(outputStream* st, const char *fmt, ...) {
+  // We only print numbers, so a small buffer is fine.
+  char buf[64];
+  va_list args;
+  int len = 0;
+  va_start(args, fmt);
+  len = jio_vsnprintf(buf, sizeof(buf), fmt, args);
+  va_end(args);
+  assert((size_t)len < sizeof(buf), "Truncation. Increase bufsize.");
+  if (st != NULL) {
+    st->print_raw(buf);
+  }
+  return len;
+}
+
+// length of time stamp
+#define TIMESTAMP_LEN 19
+// number of spaces after time stamp
+#define TIMESTAMP_DIVIDER_LEN 3
+static void print_timestamp(outputStream* st, time_t t) {
+  struct tm _tm;
+  if (os::localtime_pd(&t, &_tm) == &_tm) {
+    char buf[32];
+    ::strftime(buf, sizeof(buf), "%Y-%m-%d %H:%M:%S", &_tm);
+    st->print("%*s", TIMESTAMP_LEN, buf);
+  }
+}
+
+// A little RAII class to temporarily change locale to "C"
+class CLocaleMark {
```

Comment:
This is not threadsafe since it is a global configuration. A parallel running thread which runs locale sensitive code would be effected.